src/arithmetic/arithmetic_expression.c

```diff
@@ -135,14 +135,56 @@ int AR_EXP_GetOperandType(AR_ExpNode *exp) {
 	return -1;
 }
 
+/* Compact tree by evaluating constant expressions
+ * e.g. MINUS(X) where X is a constant number will be reduced to
+ * a single node with the value -X
+ * PLUS(MINUS(A), B) will be reduced to a single constant: B-A. */
+int AR_EXP_ReduceToScalar(AR_ExpNode **root) {
+	if((*root)->type == AR_EXP_OPERAND) {
+		if((*root)->operand.type == AR_EXP_CONSTANT) {
+			// Root is already a constant
+			return 1;
+		}
+		// Root is variadic, no way to reduce.
+		return 0;
+	} else {
+		// root represents an operation.
+		assert((*root)->type == AR_EXP_OP);
+
+		if((*root)->op.type == AR_OP_FUNC) {
+			/* See if we're able to reduce each child of root
+			 * if so we'll be able to reduce root. */
+			bool reduce_children = true;
+			for(int i = 0; i < (*root)->op.child_count; i++) {
+				if(!AR_EXP_ReduceToScalar((*root)->op.children + i)) {
+					reduce_children = false;
```

Comment:
change return type to boolean and then
`reduce_children = reduce_children && AR_EXP_ReduceToScalar((*root)->op.children + i);`

Reply:
Agreed on changing the return type, but I think this assignation is more readable (and since we break upon a failure, the number of conditionals is ultimately the same).